pkg/microservice/packager/core/service/service.go

```diff
@@ -0,0 +1,216 @@
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+
+	"github.com/koderover/zadig/pkg/microservice/packager/config"
+	"github.com/koderover/zadig/pkg/setting"
+	"github.com/koderover/zadig/pkg/tool/log"
+)
+
+// Packager ...
+type Packager struct {
+	Ctx *Context
+}
+
+type PackageResult struct {
+	ServiceName string       `json:"service_name"`
+	Result      string       `json:"result"`
+	ErrorMsg    string       `json:"error_msg"`
+	ImageData   []*ImageData `json:"image_data"`
+}
+
+func NewPackager() (*Packager, error) {
+	context, err := ioutil.ReadFile(config.JobConfigFile())
+	if err != nil {
+		return nil, err
+	}
+
+	var ctx *Context
+	if err := yaml.Unmarshal(context, &ctx); err != nil {
+		return nil, err
+	}
+
+	packager := &Packager{
+		Ctx: ctx,
+	}
+
+	return packager, nil
+}
+
+// BeforeExec ...
+func (p *Packager) BeforeExec() error {
+	log.Info("wait for docker daemon to start")
+	for i := 0; i < 120; i++ {
+		err := dockerInfo().Run()
+		if err == nil {
+			break
+		}
+		time.Sleep(time.Second * 1)
+	}
+
+	if len(p.Ctx.ProgressFile) == 0 {
+		return nil
+	}
+
+	// create log file
+	if err := os.MkdirAll(filepath.Dir(p.Ctx.ProgressFile), 0770); err != nil {
+		return errors.Wrapf(err, "failed to create progress file dir")
+	}
+	file, err := os.Create(p.Ctx.ProgressFile)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create progress file")
+	}
+	err = file.Close()
+	return errors.Wrapf(err, "failed to close progress file")
+}
+
+// Exec ...
+func (p *Packager) Exec() error {
+
+	// add docker registry auths for all registries
+	allRegistries := make([]*DockerRegistry, 0)
+	allRegistries = append(allRegistries, p.Ctx.SourceRegistries...)
+	allRegistries = append(allRegistries, p.Ctx.TargetRegistries...)
+	err := writeDockerConfig(allRegistries)
+	if err != nil {
+		return err
+	}
+
+	realTimeProgress := make([]*PackageResult, 0)
+
+	for _, image := range p.Ctx.Images {
+		cmds := p.dockerCommands(image)
+		var err error
+		for _, cmd := range cmds {
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+			log.Info(strings.Join(cmd.Args, " "))
+			if err = cmd.Run(); err != nil {
+				break
+			}
+		}
+
+		result := &PackageResult{
+			ServiceName: image.ServiceName,
+		}
+
+		if err != nil {
+			result.Result = "failed"
+			result.ErrorMsg = err.Error()
+			log.Infof("[result][fail][%s][%s]", image.ServiceName, err)
+		} else {
+			result.Result = "success"
+			result.ImageData = image.Images
+			log.Infof("[result][success][%s]", image.ServiceName)
+		}
+
+		realTimeProgress = append(realTimeProgress, result)
+
+		if len(p.Ctx.ProgressFile) > 0 {
+			bs, err := json.Marshal(realTimeProgress)
+			if err != nil {
+				log.Errorf("failed to marshal progress data %s", err)
+				continue
+			}
+			err = os.WriteFile(p.Ctx.ProgressFile, bs, 0644)
+			if err != nil {
+				log.Errorf("failed to write progress data %s", err)
+				continue
+			}
+		}
+	}
+
+	time.Sleep(time.Second * 10)
```

Comment:
what's the reason to sleep here?

Reply:
the runner of this job will catch progress info every 5 seconds, so the job need to stay alive for some time to make sure the runner can catch total progress info. The current way to export progress info is a quick interim solution, it may be replaced by a better way in a very near future.